pkg/expression/builtin.go

```diff
@@ -46,6 +46,16 @@ import (
 	"github.com/pingcap/tipb/go-tipb"
 )
 
+var _ contextopt.RequireOptionalEvalProps = notRequireOptionalEvalProps{}
+
+// notRequireOptionalEvalProps indicates the function does not require any optional properties.
+type notRequireOptionalEvalProps struct{}
+
+// RequiredOptionalEvalProps implements the RequireOptionalEvalProps interface.
+func (notRequireOptionalEvalProps) RequiredOptionalEvalProps() (set OptionalEvalPropKeySet) {
```

Comment:
What about a type cast like this

```go
type XXX interface {
    RequiredOptionalEvalProps() (set OptionalEvalPropKeySet
}
if s, ok := sig.(); ok {
    ...
}
```

If a sig didn't implement `RequiredOptionalEvalProps()`, we can assume that it does not need any `OptionalEvalPropKeySet` :thinking: . It leaves the complexity from the implementation to the caller. In my imagination, the count of callers will be much smaller than the implementation, so it'll lead to less modification. 

(But actually, this opinion is not very strong).